feat: add spring boot example which interacts with blob storage (#878)
Azure/aad-pod-identity: examples/java-blob/README.md

# Spring Boot + Azure Storage + AAD Pod Identity

This tutorial is based on [this documentation](https://docs.microsoft.com/en-us/azure/developer/java/spring-framework/configure-spring-boot-starter-java-app-with-azure-storage).

> This repository contains sample code and instructions demonstrating how you can access Azure Storage using a user assigned manage identity with the aid of AAD Pod Identity from a spring boot application.

Reviewer: ```suggestion
This repository contains sample code and instructions demonstrating how you can access Azure Storage using a user-assigned identity with the aid of AAD Pod Identity from a spring boot application.
```

> This sample takes a `POST` request and saves in as a text file in an Azure Blob Storage container.

Reviewer: ```suggestion
This sample takes a `POST` request and saves it as a text file in an Azure Blob Storage container.
```

## Prerequisites

> * AKS cluster

Reviewer: ```suggestion
* an AKS cluster with [managed identity enabled](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity)
```

* AAD Pod Identity installed and configured
* Azure Container Registry
* Azure Storage
    * Make sure you have already created a container

## Setup

> ### Create Manage Identity and Assign Role

Reviewer: ```suggestion
### Create a Managed Identity and Assign Roles
```

> In this step we'll create a new manage identity which will be used to interact with the Azure Storage Account.

Reviewer: ```suggestion
In this step, we'll create a new user-assigned identity which will be used to interact with the Azure Storage account.
```

**Note:** this step can be skipped if you already have a manage identity you'd like to reuse.

1. Setup:

    ```sh
    CLUSTER_NAME=<YOUR_AKS_CLUSTER_NAME>

>     CLUSTER_RESOURCEGROUP=$(az aks list --query "[?name == '$CLUSTER_NAME'].resourceGroup" -o tsv)

Reviewer: ```suggestion
    CLUSTER_RESOURCE_GROUP=$(az aks list --query "[?name == '$CLUSTER_NAME'].resourceGroup" -o tsv)
```

    IDENTITY_NAME=<YOUR_IDENTITY_NAME>

>     IDENTITY_RESOURCE_GROUP=<YOUR_RESOURCE_GROUP>

Reviewer: For simplicity, we can use the node resource group to store the user-assigned identity:
```suggestion
    IDENTITY_RESOURCE_GROUP=$(az aks show -g $CLUSTER_RESOURCE_GROUP -n $CLUSTER_NAME --query nodeResourceGroup -otsv)
```

    STORAGE_ACCOUNT_NAME=<STORAGE_ACCOUNT_NAME>

>     STORAGE_ACCOUNT_RESOURCEGROUP=$(az storage account list --query "[?name == '$STORAGE_ACCOUNT_NAME'].resourceGroup" -o tsv)

Reviewer: ```suggestion
    STORAGE_ACCOUNT_RESOURCE_GROUP=$(az storage account list --query "[?name == '$STORAGE_ACCOUNT_NAME'].resourceGroup" -o tsv)
```

>     CLUSTER_MSI_CLIENTID=$(az aks show \

Reviewer: ```suggestion
    CLUSTER_MSI_CLIENT_ID=$(az aks show \
```

        -n $CLUSTER_NAME \
        -g $CLUSTER_RESOURCEGROUP \
        --query "identityProfile.kubeletidentity.clientId" \
        -o tsv)

>     STORAGE_ACCOUNT_RESOURCEID=$(az storage account show \

Reviewer: ```suggestion
    STORAGE_ACCOUNT_RESOURCE_ID=$(az storage account show \
```

        --name $STORAGE_ACCOUNT_NAME \
        --query 'id' -o tsv)

>     ```

Reviewer: Could you rename the environment variables to those used in https://azure.github.io/aad-pod-identity/docs/demo/standard_walkthrough/?

Author: No problem! Done.

2. Create the manage identity:

    ```sh
    IDENTITY_RESOURCE_ID=$(az identity create \
        --name $IDENTITY_NAME \
        --resource-group $IDENTITY_RESOURCE_GROUP \
        --query 'id' -o tsv)

    IDENTITY_CLIENT_ID=$(az identity show \
        -n $IDENTITY_NAME \
        -g $IDENTITY_RESOURCE_GROUP \
        --query 'clientId' -o tsv)
    ```

> 3. Assign role to allow AAD Pod Identity to access our newly created manage identity:

Reviewer: ```suggestion
3. Assign role to allow AAD Pod Identity to access our newly created managed identity:
```

    ```sh
    az role assignment create \
        --role "Managed Identity Operator" \
        --assignee $CLUSTER_MSI_CLIENTID \
        --scope $IDENTITY_RESOURCE_ID
    ```

4. Grant permission to specific container in the Azure Storage Account:

    ```sh
    CONTAINER=test

    az role assignment create \
        --role "Storage Blob Data Contributor" \
        --assignee $IDENTITY_CLIENT_ID \

>         --scope "$(STORAGE_ACCOUNT_RESOURCEID)/blobServices/default/containers/$(CONTAINER)"

Reviewer: ```suggestion
        --scope "$STORAGE_ACCOUNT_RESOURCE_ID/blobServices/default/containers/$CONTAINER"
```

    ```

>     **Note**: if you want the manage identity to access your entire Storage Account, you can ignore `/blobServices/default/containers/$(CONTAINER)`.

Reviewer: ```suggestion
    **Note**: if you want the managed identity to access your entire Storage Account, you can ignore `/blobServices/default/containers/$(CONTAINER)`.
```

### Configure AAD Pod Identity

The following step will create a new `AzureIdentity` resource in Kubernetes in the `blob` namespace. If your namespace is something different then change NAMESPACE to match the namespace you want to deploy into.

```sh

> NAMESPACE=blob

Reviewer: ```suggestion
NAMESPACE=blob
kubectl create namespace $NAMESPACE
```

cat << EOF | kubectl apply -f -
apiVersion: "aadpodidentity.k8s.io/v1"
kind: AzureIdentity
metadata:
  name: $IDENTITY_NAME
  namespace: $NAMESPACE
spec:
  type: 0
  resourceID: $IDENTITY_RESOURCE_ID
  clientID: $IDENTITY_CLIENT_ID
EOF
```

Now we'll create the `AzureIdentityBinding` and specify the selector.

```sh

> POD_LABEL_SELECTOR=IDENTITY_NAME

Reviewer: ```suggestion
POD_LABEL_SELECTOR=$IDENTITY_NAME
```

cat << EOF | kubectl apply -f -
apiVersion: aadpodidentity.k8s.io/v1
kind: AzureIdentityBinding
metadata:

>   name: $(IDENTITY_NAME)-binding

Reviewer: ```suggestion
  name: $IDENTITY_NAME-binding
```

# Spring Boot + Azure Storage + AAD Pod Identity

This tutorial is based on [this documentation](https://docs.microsoft.com/en-us/azure/developer/java/spring-framework/configure-spring-boot-starter-java-app-with-azure-storage).

This repository contains sample code and instructions demonstrating how you can access Azure Storage using a user-assigned identity with the aid of AAD Pod Identity from a spring boot application.

This sample takes a `POST` request and saves it as a text file in an Azure Blob Storage container.

## Prerequisites

* an AKS cluster with [managed identity enabled](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity)
* AAD Pod Identity installed and configured
* Azure Container Registry
* Azure Storage
    * Make sure you have already created a container

## Setup

### Create a Managed Identity and Assign Roles

In this step, we'll create a new user-assigned identity which will be used to interact with the Azure Storage account.

> **Note:** this step can be skipped if you already have a manage identity you'd like to reuse.

Reviewer 2: nit: s/manage/managed

1. Setup:

    ```sh
    CLUSTER_NAME=<YOUR_AKS_CLUSTER_NAME>
    CLUSTER_RESOURCE_GROUP=$(az aks list --query "[?name == '$CLUSTER_NAME'].resourceGroup" -o tsv)

    IDENTITY_NAME=<YOUR_IDENTITY_NAME>
    IDENTITY_RESOURCE_GROUP=$(az aks show -g $CLUSTER_RESOURCE_GROUP -n $CLUSTER_NAME --query nodeResourceGroup -otsv)

    STORAGE_ACCOUNT_NAME=<STORAGE_ACCOUNT_NAME>
    STORAGE_ACCOUNT_RESOURCE_GROUP=$(az storage account list --query "[?name == '$STORAGE_ACCOUNT_NAME'].resourceGroup" -o tsv)

    CLUSTER_MSI_CLIENT_ID=$(az aks show \
        -n $CLUSTER_NAME \
        -g $CLUSTER_RESOURCE_GROUP \
        --query "identityProfile.kubeletidentity.clientId" \
        -o tsv)

    STORAGE_ACCOUNT_RESOURCE_ID=$(az storage account show \
        --name $STORAGE_ACCOUNT_NAME \
        --query 'id' -o tsv)
    ```

> 2. Create the manage identity:

Reviewer 2: nit: s/manage/managed
